bcycle-backend/Services/GroupTripService.cs

```diff
@@ -0,0 +1,180 @@
+            _points = context.GroupTripPoints;
+            _dbContext = context;
+        }
+
+        public async Task<GroupTrip> CreateAsync(GroupTripRequest data, string subjectId)
+        {
+            SetOrder(data.Route);
+
+            var trip = new GroupTrip
+            {
+                Name = data.Name,
+                Description = data.Description,
+                HostId = subjectId,
+                TripCode = GenerateCode(),
+                StartDate = data.StartDate,
+                EndDate = data.EndDate,
+                Route = data.Route
+            };
+
+            await _trips.AddAsync(trip);
+            await _dbContext.SaveChangesAsync();
+            return trip;
+        }
+        
+        private string GenerateCode()
+        {
+            string code;
+            do
+            {
+                code = Guid.NewGuid().ToString().Substring(0, CodeLength);
+            } while (_trips.Any(t => t.TripCode == code));
+            
+            return code;
+        }
+
+        public async Task<GroupTrip> UpdateAsync(int tripId, GroupTripRequest data, string subjectId)
+        {
+            var trip = await FindActiveAsync(tripId, subjectId);
+            if (trip == null) return null;
+
+            _points.RemoveRange(trip.Route);
+            SetOrder(data.Route);
+            trip.Update(data);
+            await _dbContext.SaveChangesAsync();
+            return trip;
+        }
+
+        private void SetOrder(List<GroupTripPoint> route)
+        {
+            for (int i = 0; i < route.Count; i++)
+            {
+                route[i].Ordinal = i;
+            }
+        }
+
+        public async Task<GroupTrip> FindAsync(int id, string subjectId)
+        {
+            var trip = await _trips
+                .Where(t => t.Id == id)
+                .Include(t => t.Participants)
+                .Include(t => t.Route)
+                .FirstOrDefaultAsync();
+
+            if (trip == null) return null;
+            if (trip.HostId == subjectId) return trip;
+
+            var participant = trip.FindAmongAccepted(subjectId);
+            if (participant == null) return null;
+
+            trip.Participants = trip.Participants.Where(p => p.Status == ParticipantStatus.Accepted).ToList();
+            return trip;
+        }
+
+        public async Task<GroupTrip> RemoveAsync(int tripId, string subjectId)
+        {
+            var trip = await _trips
+                .Where(gt => gt.Id == tripId)
+                .Where(gt => gt.HostId == subjectId)
+                .FirstOrDefaultAsync();
+
+            if (trip == null) return null;
+
+            _trips.Remove(trip);
+            await _dbContext.SaveChangesAsync();
+            return trip;
+        }
+
+        public async Task<GroupTrip> Join(string subjectId, string code)
+        {
+            var trip = await _trips
+                .Where(gt => gt.StartDate > DateTime.Now)
+                .Where(gt => gt.TripCode == code)
```

Comment:
Trip codes should be case-insensitive to protect against users typing the wrong case by mistake or because their keyboard starts sentences with an uppercase letter. (`gt.TripCode == code.ToLower()` here, `.ToLower()` when generating the codes in `GenerateCode()`.)

Reply:
Good point, but guid toString() is lowercase already

Reply:
It’s better to be explicit than to require the reader to know the exact implementation details of the `Guid` class.

Reply:
No, i dont think so. Not in this case. You're talking like it was some arcane knowledge and its a basic mechanism not even c# specific.

Reply:
Please still add `.toLower()` when searching for a code on this line, because users can still type an uppercase code.

Reply:
You mean ` .Where(gt => gt.TripCode == code.ToLower())`? That means we would allow mistyped codes. Code is shown in the application as lower case, there's no misleading at all. If someone got code that's lowercase and enters it uppercase that's his mistake and there's no point in allowing that.

Reply:
I believe that a good user interface should forgive simple mistakes, but I’m not going to fight over this.

Reply:
I knew that you will say something about "good user interface", but that wont make it one.